Mathlib/Data/MLList/AStar.lean

```diff
@@ -0,0 +1,269 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import Mathlib.Init.ZeroOne
+import Std.Data.MLList.Basic
+import Std.Data.RBMap.Alter
+import Std.Data.HashMap
+
+/-!
+# A^* search
+
+This implementation is only intended for use in meta code, rather than algorithmic analysis.
+Someone might enjoy writing a "mathematical" version.
+-/
+
+set_option autoImplicit true
+
+open Std MLList
+
+variable (m : Type u → Type u) [Monad m] [Alternative m]
+  (P V E : Type u) [Ord P] [Zero P] [Add P] [BEq V] [Hashable V]
+
+namespace AStar
+
+section GraphData
+
+/-- Representation of a path in a graph for A^* search. We cache the priority. -/
+structure Path where
+  /-- The cached priority of a path. -/
+  prio : P
+  /-- The vertex the path ends at. -/
+  to : V
+  /-- The list of edges in the path. -/
+  edges : List E
+
+/--
+Data structure encoding a graph for A^* search.
+
+This records the source and target of each directed edge,
+the edges leaving each vertex (as a lazy list),
+the weight of each edge,
+and the heuristic distance to a goal for each vertex.
+-/
+structure GraphData where
+  /-- Function assigning the source vertex to each edge. -/
+  s : E → V
+  /-- Function assigning the target vertex to each edge. -/
+  t : E → V
+  /-- Function return all edges leaving a given vertex. -/
+  nbhd : V → MLList m E
+  /-- The weight of an edge.
+  The priority of a path is the sum of its edge weights plus the heuristic of its final vertex. -/
+  weight : E → P
+  /-- The heuristic of a vertex.
+  The priority of a path is the sum of its edge weights plus the heuristic of its final vertex. -/
+  heuristic : V → P
+
+variable {m P V E}
+
+/-- Add an edge to a path. -/
+def cons (Γ : GraphData m P V E) (e : E) (p : Path P V E) : Path P V E :=
+{ prio := Γ.weight e + p.prio,
+  to := Γ.t e,
+  edges := e :: p.edges }
+
+end GraphData
+
+/--
+Data associated to a vertex during A^* search.
+We record the best seen path so far, and track which edges are `explored` and `remaining`.
+-/
+structure VertexData [Inhabited P] [Inhabited V] [Inhabited E] where
+  /-- The best so far path to the vertex. -/
+  bestPath : Path P V E
+  /-- The heuristic associated to the vertex. -/
+  heuristicPrio : P
+  /-- The sum of the edge weights of the best so far path, plus the heuristic of the vertex. -/
+  totalPrio : P
+  /-- The edges leaving the vertex which have already been explored. -/
+  explored : List E
+  /-- The remaining unexplored edges leaving the vertex. -/
+  remaining : MLList m E
+
+variable [Inhabited P] [Inhabited V] [Inhabited E]
+
+/-- This inhabited instance doesn't really make sense,
+but is needed to write `partial` functions below. -/
+instance : Inhabited (VertexData m P V E) :=
+  ⟨{ bestPath := { prio := default, to := default, edges := [] },
+     heuristicPrio := default
+     totalPrio := default
+     explored := []
+     remaining := .nil }⟩
```

Comment:
Maybe just write `deriving Inhabited` after `Path` and `VertexData` instead of doing this manually?